### PR TITLE
OADP-599: Adding functionality support bsl specific restic secret name

### PR DESCRIPTION
**Pull request by savitharaghunathan**
Summary:
* Remove default restic secret 
* Obtain restic secret from vsr/vsb
* Add validation
* Update unit tests

### CODE REVIEW

File: controllers/restic.go [hunk 1]
```diff
@@ -36,9 +36,17 @@ func (r *VolumeSnapshotBackupReconciler) CreateVSBResticSecret(log logr.Logger)
 		r.Log.Error(err, "unable to fetch PVC")
 		return false, err
 	}
+
+	// get restic secret name
+	rs := resticSecretName
```
Comment (shubham-pampattiwar) [inline controllers/restic.go:41]: So by if the user does not specify a secret name we assume `restic-secret` as the name ?

Comment (shubham-pampattiwar) [reply controllers/restic.go:41]: Just saw the operator PR, we will be creating `restic-secret` for the user, right ?

Comment (savitharaghunathan) [reply controllers/restic.go:41]: yes. Let me  ensure we are checking that in oadp PR. I dont think I am right now. Good catch

File: api/v1alpha1/volumesnapshotbackup_types.go [hunk 2]
```diff
@@ -26,6 +26,8 @@ type VolumeSnapshotBackupSpec struct {
 	VolumeSnapshotContent corev1.ObjectReference `json:"volumeSnapshotContent,omitempty"`
 	// Namespace where the Velero deployment is present
 	ProtectedNamespace string `json:"protectedNamespace,omitempty"`
+	// Restic Secret reference for given BSL
+	ResticSecretRef corev1.LocalObjectReference `json:"resticSecretRef,omitempty"`
```
[the remark this answers is not in the document]
Comment (savitharaghunathan) [reply api/v1alpha1/volumesnapshotbackup_types.go:30]: Should I change it to string type or keep it as localobjectref?

**Review by eemcmullan — APPROVED** [review]

File: controllers/restic.go (hunk 1 above)
Comment (savitharaghunathan) [reply controllers/restic.go:41]: we create `<bslname>-restic-secret` for every bsl on behalf of the user. `restic-secret` is a catch all.

Comment (dymurray) [reply controllers/restic.go:41]: Just a suggestion, might be worth naming this something like `volsync-restic-secret` just something to indicate that it's separate from the main velero restic usage

Comment (savitharaghunathan) [reply controllers/restic.go:41]: @dymurray good point. I was quite confused while switching back and forth. Would it help if I change `restic-secret` -> `volsync-restic-secret` in both places, OADP controller & vol snap controller?

Comment (savitharaghunathan) [reply controllers/restic.go]: updated `restic-secret` -> `dm-restic-secret`

File: api/v1alpha1/volumesnapshotbackup_types.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (shubham-pampattiwar) [reply api/v1alpha1/volumesnapshotbackup_types.go:30]: I am fine with localobjectref.

File: controllers/replicationdestination.go [hunk 3]
```diff
@@ -56,9 +56,9 @@ func (r *VolumeSnapshotRestoreReconciler) CreateReplicationDestination(log logr.
 func (r *VolumeSnapshotRestoreReconciler) buildReplicationDestination(replicationDestination *volsyncv1alpha1.ReplicationDestination, vsr *volsnapmoverv1alpha1.VolumeSnapshotRestore) error {
 
 	// get restic secret created by controller
-	resticSecretName := fmt.Sprintf("%s-secret", vsr.Name)
+	dmresticSecretName := fmt.Sprintf("%s-secret", vsr.Name)
```
Comment (shubham-pampattiwar) [inline controllers/replicationdestination.go:59]: I believe the restic secret name is incorrect here.

Comment (savitharaghunathan) [reply controllers/replicationdestination.go]: updated

**Review by shubham-pampattiwar — APPROVED** [review]

File: config/crd/bases/datamover.oadp.openshift.io_volumesnapshotbackups.yaml [hunk 4]
```diff
@@ -41,6 +41,14 @@ spec:
               protectedNamespace:
                 description: Namespace where the Velero deployment is present
                 type: string
+              resticSecretRef:
+                description: Restic Secret reference for given BSL
+                properties:
+                  name:
+                    description: 'Name of the referent. More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names
+                      TODO: Add other useful fields. apiVersion, kind, uid?'
```
Comment (dymurray) [inline config/crd/bases/datamover.oadp.openshift.io_volumesnapshotbackups.yaml:49]: I would remove this from the CRD description.

**Pull request merged**